python/cudf/cudf/_lib/scalar.pyx

```diff
@@ -75,6 +74,25 @@ cdef class Scalar:
 
         dtype = np.dtype(dtype)
 
+        # Caching Mechanism
+        
+        self._initialize_cache(value, dtype)
+        self.set_device_value(value, dtype)
+
+    def _initialize_cache(self, value, dtype):
+        self._host_value = value
+
+        # need to save the host dtype, because this determines
+        # the codepath for setting the device value later
+        self._host_dtype = dtype
+
+        # do not copy to device until needed
+        self._host_value_current = True
+        self._device_value_current = False
```

Comment:
I think everything would still work. Originally I had wanted to allow for the possibility of some kind of inplace operation, either on the host or device side, needing to invalidate the cache. But I can't actually think of any such cases.

I'll update this and if the need ever arises we can set it back.

Reply:
removed this and added:
`.host_value_current()` which checks if `_host_value==None`
`.device_value_current()` which checks if `self.uptr._device_uptr == NULL`

And made it so that `from_unique_ptr` initializes `_host_value` to `None`.